### ISSUE TITLE
Setting NVRAM FA4CE28D-B62F-4C99-9CC3-6815686E30F9:gpu-power-prefs - Invalid Parameter

### ISSUE THREAD

**Issue by FredWst** [acidanthera/bugtracker#928]
Hi,

I got an error with real Mac when Setting this NVRAM key gpu-power-prefs in 
FA4CE28D-B62F-4C99-9CC3-6815686E30F9

Deleted this key after creating with sudo nvram cmd in terminal works.

Here is debug trace.
[code]
ƒ00:000 00:000 OC: OpenCore DBG-059-2020-05-20 is loading in Optional mode (0/0)...
00:017 00:017 OC: Boot timestamp - 2020.05.21 07:33:21
00:036 00:018 OCCPU: MP services threads 4 (enabled 4) - Success
00:055 00:018 OCCPU: MP services Pkg 1 Cores 8 Threads 16 - Success
00:073 00:018 OCCPU: Found Intel(R) Core(TM) i7 CPU       M 640  @ 2.80GHz
00:092 00:018 OCCPU: Signature 20655 Stepping 5 Model 25 Family 6 Type 0 ExtModel 2 ExtFamily 0 uCode 7
00:111 00:018 OCCPU: Detected Apple Processor Type: 07 -> 0702
00:129 00:018 OCCPU: EIST CFG Lock 0
00:148 00:018 OCCPU: Ratio Min 9 Max 21 Current 22 Turbo 26 24 0 0
00:167 00:018 OCCPU: Timer address is 408 from LPC
00:286 00:119 OCCPU: CPUFrequencyFromTSC  2793163048Hz  2793MHz
00:306 00:019 OCCPU: CPUFrequency  2793163048Hz  2793MHz
00:324 00:018 OCCPU: FSBFrequency   133007764Hz   133MHz
00:343 00:018 OCCPU: Pkg 1 Cores 2 Threads 4
00:362 00:018 OC: OcLoadNvramSupport...
00:380 00:018 OC: Not deleting NVRAM 4D1EDE05-38C7-4A6A-9CC6-4BCCA8B38C14:UIScale, matches add
00:399 00:018 OC: Deleting NVRAM 4D1EDE05-38C7-4A6A-9CC6-4BCCA8B38C14:DefaultBackgroundColor - Not Found
00:418 00:018 OC: Not deleting NVRAM 7C436110-AB2A-4BBB-A880-FE41995C9F82:boot-args, matches add
00:437 00:018 OC: Deleting NVRAM FA4CE28D-B62F-4C99-9CC3-6815686E30F9:gpu-power-prefs - Not Found
00:460 00:023 OC: Setting NVRAM 4D1EDE05-38C7-4A6A-9CC6-4BCCA8B38C14:DefaultBackgroundColor - Success
00:479 00:018 OC: Setting NVRAM 4D1EDE05-38C7-4A6A-9CC6-4BCCA8B38C14:UIScale - ignored, exists
00:498 00:018 OC: Setting NVRAM 7C436110-AB2A-4BBB-A880-FE41995C9F82:boot-args - ignored, exists
00:516 00:018 OC: Setting NVRAM 7C436110-AB2A-4BBB-A880-FE41995C9F82:prev-lang:kbd - ignored, exists
00:535 00:018 OC: Setting NVRAM FA4CE28D-B62F-4C99-9CC3-6815686E30F9:gpu-power-prefs - Invalid Parameter
00:554 00:018 OC: Current version is DBG-059-2020-05-20
[/code]

Fred

**Comment by vit9696** [comment]
I do not think we can do much here. The log feels like the variable is either protected from writing or you attempted to write an empty variable?!

**Comment by jacksonium** [comment]
I also want to achieve the same but ran into the same issue. I tried to create a Dictionary fa4ce28d-b62f-4c99-9cc3-6815686e30f9 under NVRAM ADD, then create a child gpu-power-prefs with type DATA and <01000000> and it didnt work and it always has "Invalid Parameter"

I could do one of the following to get this nvram set:
1) I could boot MacOS via Command + S and then type: nvram fa4ce28d-b62f-4c99-9cc3-6815686e30f9:gpu-power-prefs=%01%00%00%00
2) Linux: printf "\x07\x00\x00\x00\x01\x00\x00\x00" > gpu-power-prefs-fa4ce28d-b62f-4c99-9cc3-6815686e30f9

But I want to do this in Opencore so in case nvram get zapped Opencore can put this back on prior to any OS booting.

This is important for those of us that has a defective AMD GPU on the Macbook Pro 2011. 

Is there a way to do this at all?

**Comment by jacksonium** [comment]
I also want to achieve the same but ran into the same issue. I tried to create a Dictionary fa4ce28d-b62f-4c99-9cc3-6815686e30f9 under NVRAM ADD, then create a child gpu-power-prefs with type DATA and <01000000> and it didnt work and it always has "Invalid Parameter"

I could do one of the following to get this nvram set:

I could boot MacOS via Command + S and then type: nvram fa4ce28d-b62f-4c99-9cc3-6815686e30f9:gpu-power-prefs=%01%00%00%00
Linux: printf "\x07\x00\x00\x00\x01\x00\x00\x00" > gpu-power-prefs-fa4ce28d-b62f-4c99-9cc3-6815686e30f9
But I want to do this in Opencore so in case nvram get zapped Opencore can put this back on prior to any OS booting.

This is important for those of us that has a defective AMD GPU on the Macbook Pro 2011.

Is there a way to do this at all?

**Comment by vit9696** [comment]
Probably missing Delete.

**Comment by jacksonium** [comment]
no tried delete also...

**Comment by vit9696** [comment]
Then it probably is restricted by the firmware, not much we can do here.

**Comment by jacksonium** [comment]
if it is restricted by firmware, how come under linux/cmd-S bootup in MacOS it can be set?

**Comment by vit9696** [comment]
Are you sure it actually is set instead of saying it is set, and then failing? Could you provide config.plist and the log?

**Comment by vit9696** [comment]
I will reopen the issue for it not being lost. Note, the question above still holds. Also include dmpstore of all variables from NVRAM in UEFI Shell.

**Comment by jacksonium** [comment]
[opencore-2021-02-18-171901.txt](https://github.com/acidanthera/bugtracker/files/6004666/opencore-2021-02-18-171901.txt)

[config.zip](https://github.com/acidanthera/bugtracker/files/6004669/config.zip)

**Comment by jacksonium** [comment]
I also did a nvram reset after the above dmpstore and then set the variable under linux again. 

I noticed that even if the variable is set successfully(can be seen under linux and notice there are more variable seen in linux than dmpstore ) 
![efi_linux](https://user-images.githubusercontent.com/76935685/108401463-55ac2b80-7257-11eb-8da3-8742cf6da04b.jpeg)

Doing dmpstore gpu-power-prefs under UEFI shell will return empty (vs if a variable is not present it would say so)

**Comment by vit9696** [comment]
Hmm, I am not sure what format is dmpstore file you sent me. I needed the output of dmpstore -all (check the flags, they differ slightly depending on UEFI Shell version) command in text form.

**Comment by jacksonium** [comment]
I used the uefi shell in arch linux and it said dmpstore: Unknown flag - '-all'

Is there any particular uefi shell i should use?

**Comment by jacksonium** [comment]
This is the dmpstore and ls under /sys/firmware/efi/efivars when gpu-power-prefs variable is presence:

[gpu-prefs-present-output.zip](https://github.com/acidanthera/bugtracker/files/6007224/gpu-prefs-present-output.zip)

**Comment by vit9696** [comment]
Hmmm, apparently the access to `fa4ce28d-b62f-4c99-9cc3-6815686e30f9` GUID is entirely prohibited for the bootloader (yet not the operating system). I am not sure why this happens, and I am afraid extensive firmware reverse-engineering will be needed to understand that.

**Comment by jacksonium** [comment]
Just for record, this is what it takes to do it under linux:

cd /

umount /sys/firmware/efi/efivars/

mount -t efivarfs rw /sys/firmware/efi/efivars/

cd /sys/firmware/efi/efivars/

chattr -i "gpu-power-prefs-fa4ce28d-b62f-4c99-9cc3-6815686e30f9"

rm gpu-power-prefs-fa4ce28d-b62f-4c99-9cc3-6815686e30f9

printf "\x07\x00\x00\x00\x01\x00\x00\x00" > gpu-power-prefs-fa4ce28d-b62f-4c99-9cc3-6815686e30f9

chattr +i "gpu-power-prefs-fa4ce28d-b62f-4c99-9cc3-6815686e30f9"

cd /

umount /sys/firmware/efi/efivars/

**Comment by jacksonium** [comment]
**Similarly, I can set the variable under SIP disabled and cmd-S. But it cannot be seen by nvram -p.** 

I read it from [https://wikileaks.org/ciav7p1/cms/page_26968084.html](url)

that:"...However, the manpage does not tell you everything you need to know about nvram. Although nvram -p claims to print all of the firmware variables, it does not print any of the variables that belong to the Efi GUID."

What "Efi GUID" is beyond what I can understand. I wonder if this has something to do with the problem here.

**Comment by vit9696** [comment]
`nvram -p` only prints from Apple vendor GUID. To print the value you must use `nvram fa4ce28d-b62f-4c99-9cc3-6815686e30f9:gpu-power-prefs`.

**Comment by startergo** [comment]
This is very strange. On MacBook Pro 11,3 without OC:
```
nvram fa4ce28d-b62f-4c99-9cc3-6815686e30f9:gpu-power-prefs          
fa4ce28d-b62f-4c99-9cc3-6815686e30f9:gpu-power-prefs	%00%00%00%00

```
With OC:
```
00:176 00:002 OC: Setting NVRAM FA4CE28D-B62F-4C99-9CC3-6815686E30F9:01000000 - Not Found

```

I am using [this](https://github.com/0xbb/gpu-switch) to disable DGPU, which works fine outside of OC. With OC the Nvidia DGPU is always ON. The same code does not work. The variable is set to be deleted and reset to 1 in the config file, but it is not detected in OC.

**Comment by startergo** [comment]
[This](http://dosdude1.com/apps/MacBook%20Pro%20dGPU%20Disabler.zip) Application reset the 
```fa4ce28d-b62f-4c99-9cc3-6815686e30f9:gpu-power-prefs %01%00%00%00``` from within Big Sur after second reboot.

**Comment by PMheart** [comment]
@FredWst @jacksonium @startergo I just created a new booter quirk named `DisableExternalAppleGpu`. If you are still interested, please try the artifacts at https://github.com/acidanthera/OpenCorePkg/actions/runs/2197979254.

**Comment by startergo** [comment]
> I just created a new booter quirk named `DisableExternalAppleGpu`

Thanks, but I traded-in my MBP11,3 now only have MBP11,4 (intel only).
Is this quirk only for internal Apple discrete card via multiplexer or can it be used on a desktop somehow?

**Comment by PMheart** [comment]
What is the result of `nvram fa4ce28d-b62f-4c99-9cc3-6815686e30f9:gpu-power-prefs`? Since we had invalid parameter error before, if this time it works, then it should also be fine.

And I don’t know to be honest - I just tried implementing this variable setting.

**Comment by jacksonium** [comment]
sorry what exactly do I have to do with the artifacts? with some pointer to the right direction i can give it a try.. many thanks!

**Comment by PMheart** [comment]
@jacksonium Please download the artifacts and enable the new `DisableExternalAppleGpu` quirk under `Booter->Quriks`, and see if this NVRAM key can be correctly set. Thanks.

**Comment by jacksonium** [comment]
do i need to replace the whole OC folder or can just drop in OpenCore.efi and then enable the new DisableExternalAppleGpu ?

**Comment by PMheart** [comment]
Yes, simply replacing OpenCore.efi and updating config will work.

**Comment by jacksonium** [comment]
[config.plist.zip](https://github.com/acidanthera/bugtracker/files/8532756/config.plist.zip)

Just tried and it didnt work.. could you please take a look at the attached config to make sure I made no mistakes? I basically just added DisableExternalAppleGpu and also under nvram session I added fa4ce28d-b62f-4c99-9cc3-6815686e30f9, together with replacing OpenCore.efi as you said.

**Comment by PMheart** [comment]
Yes, you did it correctly. Then it did not work for real as someone told me yesterday as well. I will take a look at it again and will see if we can do anything else with this.

**Comment by mikebeaton** [comment]
It looks plausible that EFI_VARIABLE_BOOTSERVICE_ACCESS may be denied on purpose for this variable, i.e. (I'm guessing) because it takes effect so early that it would be misleading (by implying it will work on the current boot, when it won't) to allow setting it in boot services.

I hope I understood correctly from the comments that it can be set in macOS even when not in single user mode, if SIP is disabled. (I tested setting it on an MBP10,2 with no external GPU - but not surprisingly nothing special, it can just be set, so I could not confirm.)

Sorry the first part's a bit of a guess, but if it's right there's no real mystery here: we simply cannot set it in pre-boot (nor can anyone else). Additionally, macOS (but not other OSes) is protecting writing it after boot via SIP (but this can be disabled).

I believe a working quirk in OC would have to do something like check if it is set, if it is not then exit boot services, set it, and restart immediately. And then hope it stays set, since this could render the machine unbootable if it gets in a loop.

@jacksonium: What do you have to do currently, to work round the problem "This is important for those of us that has a defective AMD GPU on the Macbook Pro 2011. Setting this variable will tell the machine to use iGPU instead of the AMD GPU.", when the variable gets unset? I mean, if it's reasonably manageable, it might not be worth such a quirk, really?

**Comment by startergo** [comment]
@mikebeaton 
```
As of Catalina, Apple has tightened EFI security of dGPU equipped MBPs and therefore gpu-power-prefs variable cannot be accessed from the EFI side. Debugged with the 2015 15" [MBP](https://egpu.io/go/macbookpro) and it returned "Access denied". This is firmware change so it will not work even if you try previous macOS versions.

MuxSetting key has no effect anymore.
```
https://egpu.io/forums/mac-setup/automate-eGPU-efi-eGPU-boot-manager-for-macos-and-windows/

**Comment by protocold** [comment]
@mikebeaton it isnt a workaround but i have to boot linux via usb stick to set this variable and after that the variable will stick around UNTIL next nvram reset. I can then bootup MacOS normally with iGPU.

For those that has defective AMD gpu, we wont be unsetting this as we just want the iGPU to take over completely.

**Comment by mikebeaton** [comment]
@vit9696 - I believe my suggested quirk would work, but is pretty horrible. Best version I can see is to reboot immediately in the middle of our exit boot services hook (maybe, only if we are booting macOS) if this var is unset, which is the soonest we can check and write it. Thoughts?!

@protocold - Written as suggested above, this would work so that you see the OC boot picker (if it's enabled/selected), regardless of whether this variable is set - does the OC picker work on these systems regardless of the setting? - and then, _only once you have selected macOS_, if this var needs setting, then the system actually immediately restarts, with the var set. Independently of whether Vit wants such hacks in his system, does that even sound like it would be useful to you?

**Comment by mikebeaton** [comment]
Oh. I've got a better one. We upgrade reset nvram to exit boot services and then set this, after doing the rest of its work, if the quirk is enabled. I think that would actually work. @vit9696 - could still do with feedback about whether I'm missing anything really horrible about trying to do this?

**Comment by mikebeaton** [comment]
Oh! Apparently I'm more or less suggesting what @PMheart already tried in that test version (at @vit9696's suggestion). So looks like that route is already closed. I'll shout if I manage to discuss it with them and suggest anything actually new - or they'll shout if they come up with anything new themselves, no doubt! - but definitely no guarantees.

**Comment by mikebeaton** [comment]
Hi, Ignore my previous comments, largely confused, I now think!

@vit9696, @PMheart -

1. I got somewhere with the brute forcing, the return code on trying to set an NVRAM variable on Apple firmware, just after ExitBootServices like this, is EFI_INVALID_PARAMETER. It applies to _all_ variables, all guids. Not a very useful return code, but it seems as if setting anything in NVRAM is locked down at this stage. (The same code sets vars fine on non-mac firmware.)
1. About the error when trying to set this variable in NVRAM in OpenCore, it is EFI_ACCESS_DENIED, and it applies to all variables in the gApplePersonalizationVariableGuid section, whatever random name you give them.
1. Personally, I haven't got anywhere with setting a value for this variable which persists until the next boot, in either macOS nor Linux. From all I've seen, I would have suggested that these variables are runtime access volatile. But people above are clearly stating that they have done this, so I'm not sure what's going on there.

The above tests are MBP10,2 firmware 429.0.0.0.0, so possibly YMMV.

**Comment by mikebeaton** [comment]
@jacksonium @FredWst @startergo @protocold - I have finally made some progress on this, at least I am actually setting this variable using OpenCore on my MBP10,2 with the latest firmware, which blocks it otherwise.

Testing would be very helpful - if anyone can try the artifacts from here https://github.com/acidanthera/OpenCorePkg/actions/runs/3315447367 and let me know the results. If it works, great, if not please could you attach an OC debug log (and either way, it would be helpful to know the Mac model and BIOS version). To test it, you need to add a new key UnprotectAppleNvram in UEFI Quirks in your config.plist and set it to true.

**Comment by mikebeaton** [comment]
cc @khronokernel - if you have time! - tyvm

**Comment by startergo** [comment]
```
nvram FA4CE28D-B62F-4C99-9CC3-6815686E30F9:gpu-power-prefs
FA4CE28D-B62F-4C99-9CC3-6815686E30F9:gpu-power-prefs    %01%00%00%00
```
On MBP11,4 with Ventura 13.0 

```
 System Firmware Version:	476.0.0.0.0
  OS Loader Version:	564.40.4~27
```
[opencore-2022-10-26-115153.txt](https://github.com/acidanthera/bugtracker/files/9869300/opencore-2022-10-26-115153.txt)

**Comment by mikebeaton** [comment]
@startergo - Thank you very much for the test.

This demonstrates that the approach to allowing the variable to be set by OpenCore works on (slightly) different firmware version.

I know you said (pm) that you no longer have the MBP11,3 with a dGPU to actually determine whether setting the variable this way (during boot with OC, using this new quirk) successfully disables the dGPU or not

So if anyone on has a machine capable of testing the above (needs to be a machine with a dGPU, and with an updated firmware version which otherwise bans OC from setting this variable), then that would be very useful.

**Comment by mikebeaton** [comment]
Are any of the original posters still reading this issue?

It would be useful to know on which models setting gpu-power-prefs (from Linux?) definitely worked to disable dGPU. Also, did it leave everything else (including Apple bootpicker?) working normally? (Ideally model + BIOS version, but just model, or model and rough date at which it worked, would do.)

**Comment by protocold** [comment]
I am but have been busy.. will try to find some time this week to give it a go.

**Comment by kambala-decapitator** [comment]
@mikebeaton I'd like to test this on MacBookPro6,2 (it has Big Sur installed via OCLP) as I also can't edit this variable neither via `nvram` call nor from linux. Do I simply need to replace my `OpenCore.efi` with that from artifacts and apply above modifications to `config.plist`? Does debug/release version matter?

**Comment by mikebeaton** [comment]
Hi. It's starting to look like this approach in its own probably isn't sufficient, and can be unsafe on some machines. More research needed.

**Comment by kambala-decapitator** [comment]
on topic of `gpu-power-prefs`, I've discovered another variable that allows selecting active GPU at boot time (well, at least on my machine):

```
nvram GfxMode=4
```

to activate iGPU at boot time and power down dGPU completely. You can also set it to 0 for iGPU, but this will not power down dGPU, i.e. both dGPU and its HD Audio device will be detectable.

see my detailed info at https://forums.macrumors.com/threads/force-2011-macbook-pro-8-2-with-failed-amd-gpu-to-always-use-intel-integrated-gpu-efi-variable-fix.2037591/post-31801562

**Comment by mikebeaton** [comment]
> on topic of `gpu-power-prefs`, I've discovered another variable that allows selecting active GPU at boot time (well, at least on my machine):
> 
> ```
> nvram GfxMode=4
> ```
> 
> to activate iGPU at boot time and power down dGPU completely. You can also set it to 0 for iGPU, but this will not power down dGPU, i.e. both dGPU and its HD Audio device will be detectable.
> 
> see my detailed info at https://forums.macrumors.com/threads/force-2011-macbook-pro-8-2-with-failed-amd-gpu-to-always-use-intel-integrated-gpu-efi-variable-fix.2037591/post-31801562

This appears to have the same effect on my MBP10,1 as forcibly setting `gpu-power-prefs`, which is to say the machine has no graphics at startup and the ROM has to be reset with a SOIC clip. Discovering that that was the effect was why I rolled back on using OC to force making it possible to set gpu-power-prefs.

**Comment by startergo** [comment]
I don't have dual GPU MacBook's anymore so I can't test it. I have one Haswell HP8570w laptop with Nvidia GPU, which runs perfectly Big Sur and I can watch Apple TV there. I was inspecting it's DSDT and discovered that the IGPU is not there. They never implemented it in the firmware. That same model came with different variants: Intel, Nvidia both. [This 8570p](https://github.com/startergo/HP8570P-OpenCore) has only Intel iGPU and when I compared the DSDT's I was surprised to see the implementation. I can share the other one's DSDT if you need it.
